### DISCUSSION TITLE
Nested Array in useFieldArray

### DISCUSSION THREAD

**Discussion by akshaykumar771** [react-hook-form/react-hook-form#3702] [General] [answered]
Hello everyone. I just started using react hook form and trying to understand most of the things. I need to implement nested inputs and I am using useFieldArray api for that purpose. 

I have a parent component where i have my main form, 1st child component where i render one input field when the user clicks on the button. This is working fine.

Now I have to render 2nd child component from 1st child component based the button click event. I am unable to render the 2nd child component on click of a button. I did not find any examples relevant to that. It would be great and really helpful if any of you could help me fix this. 

1st child component looks like this 
```
const Attributes = () => {
  const { register, setValue, getValues } = useFormContext();
  const {
    fields, append, remove
  } = useFieldArray({
    name: "attributes",
  });
  return (
    <div>
      {fields.map((field, index) => (
        <div key={field.id}>
          <label>
            Name
            <input
              name={`items[${index}].value`}
              ref={register()}
              defaultValue={field.value}
            />
          </label>
          <button type="button" onClick={() => remove(index)}>Delete</button>
          <AttributeValue index={index} {...{ register }} />
        </div>
      ))}
      <button type="button" onClick={() => append()}>Add Attribute</button>
      <button
        type="button"
        onClick={() => {
          setValue("attributes", [
            getValues().attributes,
            {
              name: "append",
              nestedArray: [{ field1: "append" }]
            }
          ]);
        }}
      >
        Append Nested
      </button>
    </div>
  );
};
```
2nd child component

```
const AttributeValue = ({ nestIndex, register }) => {
  const {
    fields, remove,
  } = useFieldArray({
    name: `attributes[${nestIndex}].values`,
  });
  return (
    <div>
      {fields.map((field, k) => (
        <div key={field.id}>
          <label>
            value
            <input
              name={`attributes[${nestIndex}].values[${k}]`}
              ref={register()}
              defaultValue={field.value}
            />
          </label>
          <button type="button" onClick={() => remove(k)}>Delete</button>
        </div>
      ))}
    </div>
  );
};
export default AttributeValue;
```
Looking forward for hints, solutions or ideas. Thanks in advance

**Comment by bluebill1049** [comment, marked as answer]
can you share a codesanbox with it?

**Reply by akshaykumar771** [reply]
Here is the link.
https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-ln34x?file=/src/fieldArray.js

**Reply by bluebill1049** [reply]
have a look at this example: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-xf574?file=/src/nestedFieldArray.js

**Reply by akshaykumar771** [reply]
I cannot render anything on the browser. It just increments the render count variable on click of a button. 

I am sorry if you have misunderstood the question. I need a button inside fields.map(()=>{}). On click of this button i should be able to render nested inputs.

At the end i want my data to look something like this. 
```
"attributes": [
      {
        "name": "Technology",
        "values": [
          "GSM",
          "UMTS CS"
        ]
      }
    ],
```

**Reply by bluebill1049** [reply]
did you see my link above which populate the field array and nested field array?

**Reply by akshaykumar771** [reply]
Yes the link does not populate anything.as i told the link only updates the render count variable and in the console i see an empty object {} as and when i click

**Reply by bluebill1049** [reply]
<img width="1020" alt="Screen Shot 2020-12-16 at 9 32 52 pm" src="https://user-images.githubusercontent.com/10513364/102337180-42c27400-3fe6-11eb-9b69-148b5b205296.png">

**Reply by bluebill1049** [reply]
note here is a bug just fixed: https://github.com/react-hook-form/react-hook-form/issues/3696 but it shouldn't impact your usage.

**Reply by akshaykumar771** [reply]
Thank you so much. I got it fixed with your hint! Cheers mate

**Reply by gauravkrp** [reply]
@bluebill1049 Did you recheck your link? It does nothing except increasing the render count by 2 !

**Reply by bluebill1049** [reply]
@gauravkrp take a look at it again, it was pointed to a local copy of react hook form.

**Reply by gauravkrp** [reply]
yay! it works now but only adds the fields once and then doesn't work!

**Reply by bluebill1049** [reply]
can you elaborate on what's doesn't work? share some steps as well, consider the question was last year.

**Reply by geekysam7** [reply]
@bluebill1049 Thanks for the example, really helpful with what I want to do. Just a suggestion, on clicking the append optional nested button, I think for the ideal behavior the `setValue` function should look like this:
```
setValue("obj.test", [
  ...fields,
  {
    name: "append",
    nestedArray: [{ field1: "append", field2: "append" }]
  }
]);
```

**Reply by geekysam7** [reply]
Found the codesandbox link which does it https://codesandbox.io/embed/react-hook-form-usefieldarray-nested-arrays-x7btr.

**Reply by varunyn** [reply]
Hi, is there any example of same with v7?

**Reply by bluebill1049** [reply]
have a look at this folder: https://github.com/react-hook-form/react-hook-form/tree/master/examples

**Reply by aspirisen** [reply]
@bluebill1049 am i right that currently if you have nested arrays and you want to use i.e. `append` for the inner array you have to add wrapper component to enable usage of `useFieldArray`? So it is not possible to operate nested array in one single react component.

**Reply by jgreens** [reply]
@bluebill1049 or @geekysam7 – I am just trying to conceptualize the example in the codesandbox for nested arrays:

Why do we need to use `setValue` when appending a nested array, instead of simply appending using parent's `append` function?

Referring to this example: https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-m8w6j?file=/src/fieldArray.js:1043-1086

**Reply by bluebill1049** [reply]
`append` is the recommended option.

**Reply by vresetnikov** [reply]
Hello, the provided example won't pass the Typescript compiler checks resulting in errors, specifically when using the append buttons. 

Could you kindly provide the exactly same example CBS but written in TS, please?

**Reply by Moshyfawn** [reply]
Consider providing a sandbox with the types you've setup instead, so we can see what throws error
